vmm/src/landlock.rs

```diff
@@ -0,0 +1,130 @@
+use std::path::PathBuf;
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub enum LandlockError {
+    /// All RulesetErrors from Landlock library are wrapped in this error
+    #[error("Error creating/adding/restricting ruleset: {0}")]
+    ManageRuleset(#[source] RulesetError),
+
+    /// Error opening path
+    #[error("Error opening path: {0}")]
+    OpenPath(#[source] IoError),
+
+    /// Invalid Landlock access
+    #[error("Invalid Landlock access: {0}")]
+    InvalidLandlockAccess(String),
+
+    /// Invalid Path
+    #[error("Invalid path")]
+    InvalidPath,
+}
+
+// https://docs.rs/landlock/latest/landlock/enum.ABI.html for more info on ABI
+static ABI: ABI = ABI::V3;
+
+pub(crate) struct LandlockAccess {
+    access: BitFlags<AccessFs>,
+}
+
+impl TryFrom<&str> for LandlockAccess {
+    type Error = LandlockError;
+
+    fn try_from(s: &str) -> Result<LandlockAccess, LandlockError> {
+        if s.is_empty() {
+            return Err(LandlockError::InvalidLandlockAccess(
+                "Access cannot be empty".to_string(),
+            ));
+        }
+
+        let mut access = BitFlags::<AccessFs>::empty();
+        for c in s.chars() {
+            match c {
+                'r' => access |= AccessFs::from_read(ABI),
+                'w' => access |= AccessFs::from_write(ABI),
+                _ => {
+                    return Err(LandlockError::InvalidLandlockAccess(
+                        format!("Invalid access: {c}").to_string(),
+                    ))
+                }
+            };
+        }
+        Ok(LandlockAccess { access })
+    }
+}
+pub struct Landlock {
+    ruleset: RulesetCreated,
+}
+
+impl Landlock {
+    pub fn new() -> Result<Landlock, LandlockError> {
+        let file_access = AccessFs::from_all(ABI);
+
+        let def_ruleset = Ruleset::default()
+            .handle_access(file_access)
+            .map_err(LandlockError::ManageRuleset)?;
+
+        // By default, rulesets are created in `BestEffort` mode. This lets Landlock
+        // to enable all the supported rules and silently ignore the unsupported ones.
+        let ruleset = def_ruleset.create().map_err(LandlockError::ManageRuleset)?;
+
+        Ok(Landlock { ruleset })
+    }
+
+    pub(crate) fn add_rule(
+        &mut self,
+        path: PathBuf,
+        access: BitFlags<AccessFs>,
+    ) -> Result<(), LandlockError> {
+        // path_beneath_rules in landlock crate handles file and directory access rules.
+        // Incoming path/s are passed to path_beneath_rules, so that we don't
+        // have to worry about the type of the path.
+        let paths = vec![path.clone()];
+        let path_beneath_rules = path_beneath_rules(paths, access);
+        self.ruleset
+            .as_mut()
+            .add_rules(path_beneath_rules)
+            .map_err(LandlockError::ManageRuleset)?;
+        Ok(())
+    }
+
+    pub(crate) fn add_rule_with_access(
+        &mut self,
+        path: PathBuf,
+        access: String,
```

Comment:
Following https://github.com/cloud-hypervisor/cloud-hypervisor/pull/6214/files#diff-af381b88eea7e5e5d9bb6626898fab0a2127ae012f87424ea44b20049f9a5319

> > You could use `access: S` where `S: AsRef<str>` to avoid all `.to_string()` calls.


> as this is a single invocation, I will leave this as is.

`add_rule_with_access()` is called 14 times: each `ApplyLandlock` implementations.

Reply:
Seems like I mis-understood your comment. I re-checked this code and realized I didn't need `to_string()` in the first place. Fixed in the latest update. Thanks for pointing it out.